rpcs3/Emu/system_progress.hpp

```diff
@@ -15,21 +27,55 @@ extern atomic_t<bool> g_system_progress_stopping;
 class scoped_progress_dialog final
 {
 	// Saved previous value
-	const char* const m_prev;
+	const char* m_prev;
+	u32 m_prev_update;
 
 public:
 	scoped_progress_dialog(const char* text) noexcept
-		: m_prev(g_progr.exchange(text ? text : ""))
 	{
+		std::tie(m_prev, m_prev_update) = g_progr.atomic_op([text = text ? text : ""](progress_dialog_string_t& progr)
+		{
+			const char* old = progr.m_text;
+			progr.m_user_count++;
+			progr.m_updates++;
+			progr.m_text = text;
+			return std::make_pair(old, progr.m_updates);
+		});
 	}
 
 	scoped_progress_dialog(const scoped_progress_dialog&) = delete;
 
 	scoped_progress_dialog& operator=(const scoped_progress_dialog&) = delete;
 
+	scoped_progress_dialog& operator=(const char* text) noexcept
+	{
+		m_prev_update = g_progr.atomic_op([this, text = text ? text : ""](progress_dialog_string_t& progr)
+		{
+			progr.m_text = text;
+			progr.m_updates++;
+
+			// Correct m_prev_update only if no other updates were made by other threads so it can be tested correctly in the destructor
+			return m_prev_update + 1 == progr.m_updates ? progr.m_updates : m_prev_update;
+		});
+
+		return *this;
+	}
+
 	~scoped_progress_dialog() noexcept
 	{
-		g_progr.release(m_prev);
+		g_progr.atomic_op([this](progress_dialog_string_t& progr)
+		{
+			if (progr.m_user_count-- == 1)
+			{
+				// Clean text only on last user
+				progr.m_text = nullptr;
```

Comment:
How do you know this was the state before?
That's pure speculation.

Reply:
Added assert.